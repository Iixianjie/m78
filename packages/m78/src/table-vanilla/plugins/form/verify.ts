import { TableCell, TableRow } from "../../types/items.js";
import { _MixinSchema } from "./schema.js";
import { _MixinBase } from "./base.js";
import { TableKey } from "../../types/base-type.js";
import {
  createVerify,
  FormRejectMeta,
  FormRejectOrValues,
  FormSchemaWithoutName,
} from "@m78/form";
import { _MixinData } from "./data.js";
import { isTruthyOrZero } from "@m78/utils";
import { throwError } from "../../../common/index.js";
import { FORM_LANG_PACK_NS, i18n } from "../../../i18n/index.js";

export interface _MixinVerify extends _MixinBase, _MixinSchema, _MixinData {}

export class _MixinVerify {
  /** 获取单元格invalid状态 */
  validCheck(cell: TableCell) {
    const { invalid } = this.getSchemas(cell.row);

    if (!invalid) return true;

    return !invalid.get(cell.column.key);
  }

  /** 初始化verify实例 */
  initVerify() {
    this.verifyInstance = createVerify({
      schemas: this.config.schemas?.length
        ? {
            schemas: this.config.schemas,
          }
        : {},
      autoVerify: false,
      languagePack: i18n.getResourceBundle(i18n.language, FORM_LANG_PACK_NS),
    });
  }

  verify() {
    return this.verifyCommon(false);
  }

  verifyUpdated() {
    return this.verifyCommon(true);
  }

  verifyRow(rowKey: TableKey) {
    const row = this.table.getRow(rowKey);
    const data = this.getFmtData(row, row.data);
    const schemas = this.getSchemas(row);

    return this.innerCheck({
      row,
      values: data,
      schemas,
    });
  }

  // verify/verifyChanged 验证通用逻辑, 逐行验证数据, 发生错误时停止并返回
  async verifyCommon(onlyUpdated: boolean): Promise<FormRejectOrValues> {
    let curError: FormRejectMeta | null = null;

    const res = await this.innerGetData(async (i, key, status) => {
      if (onlyUpdated && !status.update) return false;

      const row = this.table.getRow(key);
      const schema = this.getSchemas(row);

      const [rejects] = await this.innerCheck({
        row,
        values: i,
        schemas: schema,
      });

      // 包含错误时, 中断循环
      if (rejects) {
        curError = rejects;
        return 0;
      }
    });

    if (curError) {
      return [curError, null];
    }

    return [null, res];
  }

  // 接收处理后的values和schemas进行验证, 并更新行或单元格的错误信息, 包含错误时, 会选中并高亮首个错误单元格
  innerCheck(arg: {
    row?: TableRow | TableKey;
    // 和row二选一, 传入时, 表示cell级别验证
    cell?: TableCell;
    values: any;
    schemas: FormSchemaWithoutName;
  }): Promise<FormRejectOrValues> {
    const { row: _row, cell, values, schemas } = arg;

    let row: TableRow | undefined;

    if (cell) {
      row = cell.row;
    } else if (isTruthyOrZero(_row)) {
      row = this.table.isRowLike(_row) ? _row : this.table.getRow(_row!);
    }

    if (!row) throwError("Unable to get row");

    const verify = this.getVerify();

    let cellError = this.cellErrors.get(row.key);

    if (!cellError) {
      cellError = new Map();
      this.cellErrors.set(row.key, cellError);
    }

    return verify.staticCheck(values, schemas).then((res) => {
      const [errors] = res;

      // 需要高亮的列
      let highlightColumn: TableKey | undefined;

      const existCheck: any = {};

      // errors顺序可能与实际显示不符, 需要存储后通过当前顺序查出需要高亮的首个列
      const errColumuKeys: any = {};

      if (errors) {
        for (const e of errors) {
          // 对单元格验证和行验证采用不同的行为, cell验证仅写入对应列的错误

          if (cell) {
            // 单元格验证时, 只读取与单元格有关的第一条错误进行显示
            if (cell.column.key === e.name) {
              highlightColumn = cell.column.key;

              cellError!.set(e.name, e);
              break;
            }
          } else {
            // 每列只取第一条错误
            if (!existCheck[e.name]) {
              cellError!.set(e.name, e);
              existCheck[e.name] = true;

              errColumuKeys[e.name] = true;
            }
          }
        }

        if (!cell) {
          highlightColumn = this.context.allColumnKeys.find(
            (k) => !!errColumuKeys[k]
          );
        }
      }

      // 没有任意列被标记为错误, 单元格验证时, 清理单元格错误,  行验证时, 清理行错误
      if (!highlightColumn) {
        if (cell) {
          cellError!.delete(cell.column.key);
        } else if (!errors) {
          cellError!.clear();
        }
      }

      if (isTruthyOrZero(highlightColumn)) {
        const highlightCell =
          cell || this.table.getCell(row!.key, highlightColumn!);
        this.table.highlight(highlightCell.key);
        this.table.selectCells(highlightCell.key);
      }

      return res;
    });
  }
}
