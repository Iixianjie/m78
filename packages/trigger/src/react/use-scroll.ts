import { RefObject, useEffect, useMemo, useRef } from "react";

import { dumpFn } from "@m78/utils";
import {
  createScrollTrigger,
  type ScrollTriggerInstance,
  type ScrollTriggerOption,
  type ScrollTriggerState,
} from "../scroll/scroll.js";
import { getRefDomOrDom, useFn } from "@m78/hooks";
import { isArray } from "lodash";

/** 用于use-scroll的创建配置 */
export interface UseScrollTriggerOption
  extends Omit<ScrollTriggerOption, "target" | "handle"> {
  /** 传入要绑定的滚动元素或ref, 也可以通过useScroll返回的instance.ref绑定到dom */
  el?: HTMLElement | RefObject<any>;
  /** 滚动时触发 */
  onScroll?(meta: ScrollTriggerState): void;
}

/** use-scroll扩展后的ScrollTrigger实例 */
export interface UseScrollTriggerInstance<ElType extends HTMLElement>
  extends ScrollTriggerInstance {
  /** 可使用此项代替option.el进行绑定 */
  ref: RefObject<ElType>;
}

export function useScroll<ElType extends HTMLElement>(
  option: UseScrollTriggerOption
) {
  // 用于返回的节点获取ref
  const ref = useRef<ElType>(null);

  const instance = useMemo(() => {
    return {
      ref,
      scroll: dumpFn,
      scrollToE: dumpFn,
      scrollToElement: dumpFn,
      get: dumpFn,
      destroy: dumpFn,
    } as any as UseScrollTriggerInstance<ElType>;
  }, []);

  const handle = useFn((e: ScrollTriggerState) => {
    option.onScroll?.(e);
  });

  // 在满足条件时重载实例
  const getDeps = () => {
    const _offset = option.offset;
    const offset = isArray(_offset) ? _offset : [_offset, _offset];

    const _touchOffset = option.touchOffset;
    const touchOffset = isArray(_touchOffset)
      ? _touchOffset
      : [_touchOffset, _touchOffset];

    return [
      option.el,
      ref.current,
      (option.el as any)?.current,
      offset[0],
      offset[1],
      touchOffset[0],
      touchOffset[1],
    ];
  };

  useEffect(() => {
    const curEl = getRefDomOrDom(option.el, ref);

    if (curEl) {
      Object.assign(
        instance,
        createScrollTrigger({
          ...option,
          target: curEl,
          handle,
        })
      );
    }

    return () => {
      instance.destroy?.();
    };
  }, getDeps());

  return instance;
}
