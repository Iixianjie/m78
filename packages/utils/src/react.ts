import { AnyFunction } from "./types";

/**
 * 便捷的按键和点击事件绑定
 * @param handle - 时间处理函数
 * @param spaceTrigger - 按下空格时是否触发
 * */
export function keypressAndClick(handle: AnyFunction, spaceTrigger = true) {
  return {
    onClick: handle,
    onKeyPress: (e: any) => {
      const code = e.code;

      if (code === "Enter" || (spaceTrigger && code === "Space")) {
        handle?.();
      }
    },
  };
}
