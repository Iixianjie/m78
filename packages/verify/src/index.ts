import _defaultsDeep from "lodash/defaultsDeep.js";
import { Config, Verify } from "./types.js";
import { defaultConfig } from "./default-config.js";
import { getCheckApi } from "./check.js";
import { simplifiedChinese, english } from "./language-pack.js";

function createVerify(config?: Config): Verify {
  const conf = _defaultsDeep(
    { languagePack: config?.extendLanguagePack },
    config,
    defaultConfig
  ) as Required<Config>;

  const verify: Verify = {
    languagePack: conf.languagePack,
  } as Verify;

  return Object.assign(verify, getCheckApi(conf, verify));
}

export { simplifiedChinese, english, createVerify };

export * from "./types.js";
export * from "./error.js";
export * from "./validator/required.js";
export * from "./validator/object.js";
export * from "./validator/bool.js";
export * from "./validator/string.js";
export * from "./validator/array.js";
export * from "./validator/fn.js";
export * from "./validator/number.js";
export * from "./validator/symbol.js";
export * from "./validator/regexp.js";
export * from "./validator/regexp-string.js";
export * from "./validator/pattern.js";
export * from "./validator/specific.js";
export * from "./validator/equality.js";
export * from "./validator/within.js";
export * from "./validator/without.js";
export * from "./validator/url.js";
export * from "./validator/email.js";
export * from "./validator/date.js";
export * from "./validator/match.js";
export * from "./validator/list.js";
