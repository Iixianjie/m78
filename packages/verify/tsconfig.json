{
  "compilerOptions": {
    "baseUrl": "./",
    "paths": {
    }
  },
  "extends": "@m78/build-tools/tsconfig-base.json",
  "include": [
    "src/**/*",
    "examples/**/*",
    "test/**/*"
  ]
}
