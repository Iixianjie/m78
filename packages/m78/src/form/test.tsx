import { createForm } from "./index.js";
import { required } from "@m78/verify";
import { Input } from "../input/index.js";
import React from "react";
import { render } from "@testing-library/react";

describe("form", () => {
  test("schema", async () => {
    const form = createForm({
      defaultValue: {
        name: "m78",
        desc: "",
      },
      schemas: {
        schema: [
          {
            name: "name",
            validator: required(),
            label: "姓名",
            dynamic: (form) => ({
              hidden: form.getValue("desc") === "abc",
            }),
            deps: ["desc"],
            describe:
              "填入你的姓名, 填入你的姓名填入你的姓名填入你的姓名填入你的姓名填入你的姓名, 填入你的姓名.",
            component: <Input />,
          },
          {
            name: "desc",
            label: "描述",
            validator: required(),
            component: <Input textArea />,
          },
          {
            name: "obj",
            label: "对象",
            validator: required(),
            schema: [
              {
                name: "title",
                validator: required(),
                component: <Input placeholder="名称" />,
              },
              {
                name: "desc",
                validator: required(),
                component: <Input placeholder="描述" />,
              },
            ],
          },
          {
            name: "list",
            label: "列表",
            list: true,
            dynamic: (form) => ({
              valid: form.getValue("name") !== "123",
            }),
            deps: ["name"],
            validator: required(),
            eachSchema: {
              schema: [
                {
                  name: "title",
                  validator: required(),
                  component: <Input placeholder="名称" />,
                },
                {
                  name: "desc",
                  validator: required(),
                  component: <Input placeholder="描述" />,
                },
              ],
            },
          },
          {
            name: "list2",
            label: "列表3",
            list: true,
            listDefaultValue: "11",
            dynamic: (form) => ({
              valid: form.getValue("name") !== "123",
            }),
            deps: ["name"],
            validator: required(),
            eachSchema: {
              validator: required(),
              component: <Input />,
            },
          },
        ],
      },
    });

    const r = render(<form.SchemaRender />);

    expect(r.container).toMatchSnapshot();
  });
});
