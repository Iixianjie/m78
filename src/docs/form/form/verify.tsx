import React from 'react';
import { required, string, useForm } from 'm78/form';
import { Input } from 'm78/input';

const Verify = () => {
  const Form = useForm();

  return (
    <div>
      <Form.Field name="user" label="用户名" validator={[required(), string({ min: 2, max: 6 })]}>
        <Input placeholder="输入用户名" />
      </Form.Field>
    </div>
  );
};

export default Verify;
