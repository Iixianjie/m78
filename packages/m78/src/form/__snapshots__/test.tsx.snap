// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form schema 1`] = `
<div>
  <div
    class="m78 m78-cell"
  >
    <div
      class="m78-cell_cont m78-flex-wrap"
      style="margin: 0px;"
    >
      <div
        class="m78 m78-cell_col"
        style="width: 100%; padding: 0px;"
      >
        <div
          class="m78 m78-row m78-tile m78-lay m78-form_field m78-form_horizontal __none m78-cross-start"
          style="max-width: 440px;"
        >
          <div
            class="m78-tile_leading"
          >
            <div
              class="m78-form_label m78-form_horizontal-label"
            >
              <span
                class="color-red fs-md vm mr"
              >
                *
              </span>
              姓名
            </div>
          </div>
          <div
            class="m78-tile_main"
          >
            <div
              class="m78-tile_cont"
            >
              <div
                class="m78-form_unit"
              >
                <span
                  class="m78 m78-init m78-input"
                >
                  <input
                    class="m78-input_inner"
                    type="text"
                    value="m78"
                  />
                </span>
                <div
                  class="m78-form_bubble-error"
                  style="opacity: 0; visibility: hidden;"
                />
              </div>
              <div
                class="m78-form_describe"
              >
                填入你的姓名, 填入你的姓名填入你的姓名填入你的姓名填入你的姓名填入你的姓名, 填入你的姓名.
              </div>
              <div
                class="m78-form_error"
                role="alert"
                style="opacity: 0; visibility: hidden;"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="m78 m78-cell_col"
        style="width: 100%; padding: 0px;"
      >
        <div
          class="m78 m78-row m78-tile m78-lay m78-form_field m78-form_horizontal __none m78-cross-start"
          style="max-width: 440px;"
        >
          <div
            class="m78-tile_leading"
          >
            <div
              class="m78-form_label m78-form_horizontal-label"
            >
              <span
                class="color-red fs-md vm mr"
              >
                *
              </span>
              描述
            </div>
          </div>
          <div
            class="m78-tile_main"
          >
            <div
              class="m78-tile_cont"
            >
              <div
                class="m78-form_unit"
              >
                <span
                  class="m78 m78-init m78-input __textarea"
                >
                  <textarea
                    class="m78-input_inner"
                    style="height: 0px; overflow: hidden; resize: none;"
                    type="text"
                  />
                  <textarea
                    class="m78-input_inner"
                    style="height: 0px; overflow: hidden; resize: none; position: absolute; visibility: hidden;"
                    type="text"
                  />
                </span>
                <div
                  class="m78-form_bubble-error"
                  style="opacity: 0; visibility: hidden;"
                />
              </div>
              <div
                class="m78-form_error"
                role="alert"
                style="opacity: 0; visibility: hidden;"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="m78 m78-cell_col"
        style="width: 100%; padding: 0px;"
      >
        <div
          class="m78 m78-row m78-tile m78-lay m78-form_field m78-form_horizontal __none m78-cross-start"
          style="max-width: 440px;"
        >
          <div
            class="m78-tile_leading"
          >
            <div
              class="m78-form_label m78-form_horizontal-label"
            >
              <span
                class="color-red fs-md vm mr"
              >
                *
              </span>
              对象
            </div>
          </div>
          <div
            class="m78-tile_main"
          >
            <div
              class="m78-tile_cont"
            >
              <div
                class="m78-form_unit"
              >
                <div
                  class="m78 m78-row m78-cross-start"
                >
                  <div
                    class="m78-form_multi-column"
                  >
                    <div
                      class="m78 m78-row m78-tile m78-lay m78-form_field m78-form_horizontal __none m78-cross-start"
                      style="max-width: 440px;"
                    >
                      <div
                        class="m78-tile_main"
                      >
                        <div
                          class="m78-tile_cont"
                        >
                          <div
                            class="m78-form_unit"
                          >
                            <span
                              class="m78 m78-init m78-input"
                            >
                              <input
                                class="m78-input_inner"
                                placeholder="名称"
                                type="text"
                                value=""
                              />
                            </span>
                            <div
                              class="m78-form_bubble-error"
                              style="opacity: 0; visibility: hidden;"
                            />
                          </div>
                          <div
                            class="m78-form_error"
                            role="alert"
                            style="opacity: 0; visibility: hidden;"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="m78-form_multi-column"
                  >
                    <div
                      class="m78 m78-row m78-tile m78-lay m78-form_field m78-form_horizontal __none m78-cross-start"
                      style="max-width: 440px;"
                    >
                      <div
                        class="m78-tile_main"
                      >
                        <div
                          class="m78-tile_cont"
                        >
                          <div
                            class="m78-form_unit"
                          >
                            <span
                              class="m78 m78-init m78-input"
                            >
                              <input
                                class="m78-input_inner"
                                placeholder="描述"
                                type="text"
                                value=""
                              />
                            </span>
                            <div
                              class="m78-form_bubble-error"
                              style="opacity: 0; visibility: hidden;"
                            />
                          </div>
                          <div
                            class="m78-form_error"
                            role="alert"
                            style="opacity: 0; visibility: hidden;"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="m78-form_bubble-error"
                  style="opacity: 0; visibility: hidden;"
                />
              </div>
              <div
                class="m78-form_error m78-form_empty-hide"
                role="alert"
                style="opacity: 0; visibility: hidden;"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="m78 m78-cell_col"
        style="width: 100%; padding: 0px;"
      >
        <div
          class="m78 m78-row m78-tile m78-lay m78-form_field m78-form_horizontal __none m78-cross-start"
          style="max-width: 440px;"
        >
          <div
            class="m78-tile_leading"
          >
            <div
              class="m78-form_label m78-form_horizontal-label"
            >
              <span
                class="color-red fs-md vm mr"
              >
                *
              </span>
              列表
            </div>
          </div>
          <div
            class="m78-tile_main"
          >
            <div
              class="m78-tile_cont"
            >
              <div
                class="m78-form_unit"
              >
                <div
                  class="m78-form_list-wrap"
                >
                  <div>
                    <button
                      class="m78 m78-init m78-btn m78-effect __md"
                      type="button"
                    >
                      add item
                      <svg
                        class="m78 m78-icon m78-icon_icon-add-circle-outline"
                        data-name="icon-add-circle-outline"
                        focusable="false"
                        height="24"
                        style="margin-left: 4px;"
                        viewBox="0 0 24 24"
                        width="24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 7c-.55 0-1 .45-1 1v3H8c-.55 0-1 .45-1 1s.45 1 1 1h3v3c0 .55.45 1 1 1s1-.45 1-1v-3h3c.55 0 1-.45 1-1s-.45-1-1-1h-3V8c0-.55-.45-1-1-1zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
                <div
                  class="m78-form_bubble-error"
                  style="opacity: 0; visibility: hidden;"
                />
              </div>
              <div
                class="m78-form_error"
                role="alert"
                style="opacity: 0; visibility: hidden;"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="m78 m78-cell_col"
        style="width: 100%; padding: 0px;"
      >
        <div
          class="m78 m78-row m78-tile m78-lay m78-form_field m78-form_horizontal __none m78-cross-start"
          style="max-width: 440px;"
        >
          <div
            class="m78-tile_leading"
          >
            <div
              class="m78-form_label m78-form_horizontal-label"
            >
              <span
                class="color-red fs-md vm mr"
              >
                *
              </span>
              列表3
            </div>
          </div>
          <div
            class="m78-tile_main"
          >
            <div
              class="m78-tile_cont"
            >
              <div
                class="m78-form_unit"
              >
                <div
                  class="m78-form_list-wrap"
                >
                  <div>
                    <button
                      class="m78 m78-init m78-btn m78-effect __md"
                      type="button"
                    >
                      add item
                      <svg
                        class="m78 m78-icon m78-icon_icon-add-circle-outline"
                        data-name="icon-add-circle-outline"
                        focusable="false"
                        height="24"
                        style="margin-left: 4px;"
                        viewBox="0 0 24 24"
                        width="24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 7c-.55 0-1 .45-1 1v3H8c-.55 0-1 .45-1 1s.45 1 1 1h3v3c0 .55.45 1 1 1s1-.45 1-1v-3h3c.55 0 1-.45 1-1s-.45-1-1-1h-3V8c0-.55-.45-1-1-1zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
                <div
                  class="m78-form_bubble-error"
                  style="opacity: 0; visibility: hidden;"
                />
              </div>
              <div
                class="m78-form_error"
                role="alert"
                style="opacity: 0; visibility: hidden;"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="m78 m78-cell_col"
        style="width: 100%; padding: 0px;"
      >
        <div
          class="m78 m78-row m78-tile m78-lay m78-form_field m78-form_horizontal __none m78-cross-start"
          style="max-width: 440px;"
        >
          <div
            class="m78-tile_leading"
          >
            <div
              class="m78-form_label m78-form_horizontal-label"
            >
               
            </div>
          </div>
          <div
            class="m78-tile_main"
          >
            <div
              class="m78-tile_cont"
            >
              <div
                class="m78-form_unit"
              >
                <div>
                  <button
                    class="m78 m78-init m78-btn m78-effect __md"
                    type="button"
                  >
                    reset
                  </button>
                  <button
                    class="m78 m78-init m78-btn m78-effect __md __primary __light"
                    style="min-width: 80px;"
                    type="button"
                  >
                    submit
                  </button>
                </div>
                <div
                  class="m78-form_bubble-error"
                  style="opacity: 0; visibility: hidden;"
                />
              </div>
              <div
                class="m78-form_error"
                role="alert"
                style="opacity: 0; visibility: hidden;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
